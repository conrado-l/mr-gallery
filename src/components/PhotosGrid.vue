<template>
  <div>
    <!-- Coolbox photo library -->
    <CoolLightBox
      :items="photos"
      :index="currentOpenedPhotoIndex"
      :useZoomBar="true"
      :fullScreen="true"
      :loop="false"
      :gallery="false"
      effect="fade"
      @close="currentOpenedPhotoIndex = null"
      @on-change="onPhotoDetailChange($event)"
    >
    </CoolLightBox>
    <!---->

    <!-- Infinity scroll -->
    <div v-infinite-scroll="onInfiniteScrollLoadMore"
         :infinite-scroll-distance="10">
      <v-row>
        <v-col
          v-for="(photo, photoIndex) in photos"
          :key="photo.id"
          class="d-flex child-flex"
          xs="1"
          sm="1"
          md="4"
        >
          <!-- Thumbnail photos -->
          <v-tooltip bottom>
            <template v-slot:activator="{ on, attrs }">
              <v-img
                class="grey lighten-2 clickable"
                :src="photo.thumb"
                v-ripple
                v-bind="attrs"
                v-on="on"
                data-test="thumbnail-photo"
                @click="onPhotoClick(photoIndex, photo.id)"
              >
                <template v-slot:placeholder>
                  <v-row
                    class="fill-height ma-0"
                    align="center"
                    justify="center"
                  >
                    <v-progress-circular
                      indeterminate
                      color="grey lighten-5"
                    ></v-progress-circular>
                  </v-row>
                </template>
              </v-img>
            </template>
            <span>Click for more details</span>
          </v-tooltip>
          <!---->
        </v-col>
      </v-row>
    </div>
    <!----->
  </div>
</template>

<script>
// Plugins
import CoolLightBox from 'vue-cool-lightbox'
import 'vue-cool-lightbox/dist/vue-cool-lightbox.min.css'
import infiniteScroll from 'vue-infinite-scroll'

/**
 * A grid of clickable photos, using infinity scroll, with the ability to open them in full screen, for more details
 **/
export default {

  name: 'PhotosGrid',
  props: {
    /** Contains the photos to be displayed **/
    photos: {
      type: Array,
      required: true
    },
    /** Checks if the photos are being fetched **/
    fetchingPhotos: {
      type: Boolean,
      required: true
    },
    /** Checks if the photo's details are being fetched **/
    fetchingPhotoDetails: {
      type: Boolean,
      required: true
    }
  },
  data: function () {
    return {
      currentOpenedPhotoIndex: null
    }
  },
  components: {
    CoolLightBox
  },
  directives: { infiniteScroll },
  methods: {
    /**
     * Notifies that the user scrolled down for loading more images
     /**/
    onInfiniteScrollLoadMore () {
      if (this.fetchingPhotos) {
        return
      }

      this.$emit('load-more-photos')
    },
    /**
     * Notifies that the user changed the photo from the detail modal
     * @param {number} photoIndex
     **/
    onPhotoDetailChange (photoIndex) {
      if (this.fetchingPhotoDetails) {
        return
      }

      this.notifyPhotoDetail(this.photos[photoIndex].id)
    },
    /**
     * Notifies that a photo was clicked, it also opens the detail modal for showing the photo in full size
     * @param {number} photoIndex
     * @param {string} photoId
     **/
    onPhotoClick (photoIndex, photoId) {
      this.notifyPhotoDetail(photoId)
      this.currentOpenedPhotoIndex = photoIndex
    },
    /**
     * Notifies that we need the details for a particular photo
     * @param {string} photoId
     */
    notifyPhotoDetail (photoId) {
      this.$emit('load-photo-detail', photoId)
    }
  }
}
</script>

<style lang="scss" scoped>
.clickable {
  cursor: pointer;
}
</style>
