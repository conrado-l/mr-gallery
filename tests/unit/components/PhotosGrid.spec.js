// Vue Test Utils
import { createLocalVue, mount } from '@vue/test-utils'

// Components
import PhotosGrid from '@/components/PhotosGrid'

// Mocks
import photosMock from './mocks/photos'

describe('PhotosGrid.vue', () => {
  const localVue = createLocalVue()

  // Default props
  const defaultProps = {
    photos: [],
    fetchingPhotos: false,
    fetchingPhotoDetails: false
  }

  // Used selectors
  const selectors = {
    thumbnailPhoto: '[data-test="thumbnail-photo"]',
    viewer: '[data-test="viewer-container"]'
  }

  // Reusable factory mount
  const factoryMount = (props) => {
    return mount(PhotosGrid, {
      localVue,
      propsData: {
        ...props
      }
    })
  }

  it('should render the component correctly', () => {
    const wrapper = factoryMount(defaultProps)

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render no thumbnail photos', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the thumbnail photos
    const thumbnailPhotoWrappers = wrapper.findAll(selectors.thumbnailPhoto)

    // Check if no thumbnail photos were rendered
    expect(thumbnailPhotoWrappers.length).toEqual(0)
  })

  it('should render 10 thumbnail photos', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the thumbnail photos
    const thumbnailPhotoWrappers = wrapper.findAll(selectors.thumbnailPhoto)

    // Check if 10 thumbnail photos were rendered
    expect(thumbnailPhotoWrappers.length).toEqual(10)
  })

  // TODO: do (more) research on why "vue-infinite-scroll" is not triggering in tests. The component was tested isolated, as root
  //  component, and "onInfiniteScrollLoadMore" was executed properly in the browser.
  it.skip('should initially notify that the first photos should be loaded', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps
    })

    // Mock/stub method
    const mockOnInfiniteScrollLoadMore = jest.spyOn(wrapper.vm, 'onInfiniteScrollLoadMore')

    // Check that the function has been called
    expect(mockOnInfiniteScrollLoadMore).toHaveBeenCalled()

    // Wait until $emits have been handled
    await wrapper.vm.$nextTick()

    // Check that the event was emitted
    expect(wrapper.emitted('load-more-photos')).toBeTruthy()
  })

  it('should render the thumbnail photos with the correct source URLs', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the thumbnail photos
    const thumbnailPhotoWrappers = wrapper.findAll(selectors.thumbnailPhoto)

    // Check if the thumbnail photos sources are correct (VImg uses the "background-image" CSS property)
    thumbnailPhotoWrappers.wrappers.forEach((thumbnailPhoto, index) => {
      expect(thumbnailPhoto.element.src).toBe(photosMock[index].thumbnailPhoto)
    })
  })

  it('should open the details modals when clicking on a photo', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the thumbnail photo
    const thumbnailPhotoWrapper = wrapper.find(selectors.thumbnailPhoto)

    // Click over the photo
    await thumbnailPhotoWrapper.trigger('click')

    // Check if the modal is visible
    expect(wrapper.find(selectors.viewer).exists()).toBe(true)
  })

  it('should notify to load the photo detail when clicking on a photo from the grid', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock
    })

    // Find the thumbnail photo
    const thumbnailPhotoWrapper = wrapper.find(selectors.thumbnailPhoto)

    // Click over the photo
    await thumbnailPhotoWrapper.trigger('click')

    // Wait until $emits have been handled
    await wrapper.vm.$nextTick()

    // Check that the event was emitted
    expect(wrapper.emitted('load-photo-detail')).toBeTruthy()
  })
})
