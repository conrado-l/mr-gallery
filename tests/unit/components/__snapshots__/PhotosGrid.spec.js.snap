// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotosGrid.vue should render the component correctly 1`] = `
<div
  class="d-flex-center"
>
  <transition-stub
    name="fade"
  >
    <!---->
  </transition-stub>
   
  <div>
    <div
      class="grid-container"
    />
  </div>
</div>
`;
